src/set_theory/ordinal/notation.lean

```diff
@@ -791,6 +791,142 @@ begin
+        lt_def, mul_one, nat.cast_zero, nat.cast_succ, nat.succ_pnat_coe, opow_succ,
+        opow_zero, ordinal.mul_add_one, pnat.one_coe, succ_zero, true_and, _root_.zero_add,
+        zero_def],
+    { apply_instance },
+    { exact ⟨rfl, λ _, by apply_instance⟩ },
+    { have := opow_pos _ omega_pos,
+      refine ⟨mul_is_limit this omega_is_limit,
+        λ i, ⟨this, _, λ H, @NF.oadd_zero _ _ (iha.2 H.fst)⟩, exists_lt_mul_omega'⟩,
+      rw [← mul_succ, nat_cast_succ, ordinal.mul_lt_mul_iff_left this],
+      apply nat_lt_omega },
+    { have := opow_pos _ omega_pos,
+      refine ⟨
+        add_is_limit _ (mul_is_limit this omega_is_limit), λ i, ⟨this, _, _⟩,
+        exists_lt_add exists_lt_mul_omega' _⟩,
+      { rw [← mul_succ, nat_cast_succ, ordinal.mul_lt_mul_iff_left this],
+        apply nat_lt_omega },
+      { refine λ H, H.fst.oadd _ (NF.below_of_lt' _ (@NF.oadd_zero _ _ (iha.2 H.fst))),
+        rw [repr, repr, add_zero, iha.1, opow_succ, ordinal.mul_lt_mul_iff_left this],
+        apply nat_lt_omega } },
+    { rcases iha with ⟨h1, h2, h3⟩,
+      refine ⟨opow_is_limit one_lt_omega h1, λ i, _, exists_lt_omega_opow' one_lt_omega h1 h3⟩,
+      obtain ⟨h4, h5, h6⟩ := h2 i, exact ⟨h4, h5, λ H, @NF.oadd_zero _ _ (h6 H.fst)⟩ },
+    { rcases iha with ⟨h1, h2, h3⟩,
+      refine ⟨add_is_limit _ (opow_is_limit one_lt_omega h1), λ i, _,
+        exists_lt_add (exists_lt_omega_opow' one_lt_omega h1 h3) _⟩,
+      obtain ⟨h4, h5, h6⟩ := h2 i,
+      refine ⟨h4, h5, λ H, H.fst.oadd _ (NF.below_of_lt' _ (@NF.oadd_zero _ _ (h6 H.fst)))⟩,
+      rwa [repr, repr, add_zero, coe_coe, pnat.one_coe, nat.cast_one, mul_one,
+        opow_lt_opow_iff_right one_lt_omega] } },
+  { refine ⟨by rw [repr, ihb.1, add_succ, repr],
+      λ H, H.fst.oadd _ (NF.below_of_lt' _ (ihb.2 H.snd))⟩,
+    have := H.snd'.repr_lt, rw ihb.1 at this,
+    exact (lt_succ_self _).trans this },
+  { rcases ihb with ⟨h1, h2, h3⟩,
+    simp only [repr],
+    exact ⟨ordinal.add_is_limit _ h1,
+      λ i, ⟨oadd_lt_oadd_3 (h2 i).1, oadd_lt_oadd_3 (h2 i).2.1, λ H, H.fst.oadd _
+        (NF.below_of_lt' (lt_trans (h2 i).2.1 H.snd'.repr_lt) ((h2 i).2.2 H.snd))⟩,
+      exists_lt_add h3 _⟩ }
+end
+
+/-- The fast growing hierarchy for ordinal notations `< ε₀`. This is a sequence of
+functions `ℕ → ℕ` indexed by ordinals, with the definition:
+* `f_0(n) = n + 1`
+* `f_(α+1)(n) = f_α^[n](n)`
+* `f_α(n) = f_(α[n])(n)` where `α` is a limit ordinal
+   and `α[i]` is the fundamental sequence converging to `α` -/
+def fast_growing : onote → ℕ → ℕ
+| o :=
+  match fundamental_seq o, fundamental_seq_has_prop o with
+  | sum.inl none, _ := nat.succ
+  | sum.inl (some a), h :=
+    have a < o, { rw [lt_def, h.1], apply lt_succ_self },
+    λ i, (fast_growing a)^[i] i
+  | sum.inr f, h := λ i, have f i < o, from (h.2.1 i).2.1, fast_growing (f i) i
+  end
+using_well_founded
+{ rel_tac := λ _ _, `[exact ⟨_, inv_image.wf repr ordinal.wf⟩],
```

Comment:
Perhaps you could define this as the canonical well-founded relation on `onote`.